src/client/Microsoft.Identity.Client/Internal/Requests/OnBehalfOfRequest.cs

```diff
@@ -86,18 +87,34 @@ protected override async Task<AuthenticationResult> ExecuteAsync(CancellationTok
             // No AT in the cache or AT needs to be refreshed
             try
             {
-                using (logger.LogBlockDuration("[OBO request] Fetching OBO token from ESTS"))
-                {
-                    var result = await FetchNewAccessTokenAsync(cancellationToken).ConfigureAwait(false);
-                    return result;
-                }
+                return await RefreshRtOrFetchNewAccessTokenAsync(cancellationToken).ConfigureAwait(false);
             }
             catch (MsalServiceException e)
             {
                 return HandleTokenRefreshError(e, msalAccessTokenItem);
             }
         }
 
+        private async Task<AuthenticationResult> RefreshRtOrFetchNewAccessTokenAsync(CancellationToken cancellationToken)
+        {
+            // Look for a refresh token
+            MsalRefreshTokenCacheItem appRefreshToken = await CacheManager.FindRefreshTokenAsync().ConfigureAwait(false);
+
+            // If a refresh token is not found, fetch a new access token
+            if (appRefreshToken == null)
+            {
+                AuthenticationRequestParameters.RequestContext.Logger.Verbose("[OBO request] No Refresh Token was found in the cache. Fetching OBO token from ESTS");
+
+                var result = await FetchNewAccessTokenAsync(cancellationToken).ConfigureAwait(false);
+                return result;
+            }
+
+            var msalTokenResponse = await SilentRequestHelper.RefreshAccessTokenAsync(appRefreshToken, cancellationToken, this, AuthenticationRequestParameters)
+                .ConfigureAwait(false);
+            
+            return await CacheTokenResponseAndCreateAuthenticationResultAsync(msalTokenResponse).ConfigureAwait(false);
+        }
+
         private async Task<AuthenticationResult> FetchNewAccessTokenAsync(CancellationToken cancellationToken)
```

Comment:
>FetchNewAccessTokenAsync [](http://example.com/codeflow?start=49&length=24)

This will also fetch the RT? Method name no longer accurate?

Reply:
There are no changes to this method. This is simply performing OBO to fetch a new token instead of using AT or RT from cache.